itou/utils/password_validation.py

```diff
@@ -0,0 +1,50 @@
+import string
+
+from django.core.exceptions import ValidationError
+from django.utils.translation import gettext_lazy
+
+
+class CnilCompositionPasswordValidator:
+    """
+    Validate whether the password is conform to CNIL guidelines.
+
+    CNIL guidelines regarding the use case "Avec restriction d'accès":
+    https://www.cnil.fr/fr/authentification-par-mot-de-passe-les-mesures-de-securite-elementaires
+
+    - Minimum length: at least 8 characters (handled via `MinimumLengthValidator`)
+    - Composition: password must include 3 of the 4 types of characters
+        1) uppercase
+        2) lowercase
+        3) numbers
+        4) special characters
+    - Additional measure: account access timeout after multiple failures (handled via `django-allauth`)
+
+    This validator only checks for the "composition" part.
+    """
+
+    SPECIAL_CHARS = string.punctuation
+
+    ERROR_MSG = gettext_lazy(
+        "Le mot de passe doit contenir au moins 3 des 4 types suivants : "
+        "majuscules, minuscules, chiffres, caractères spéciaux."
+    )
+
+    HELP_MSG = gettext_lazy(
+        "Votre mot de passe doit contenir au moins 3 des 4 types suivants : "
+        "majuscules, minuscules, chiffres, caractères spéciaux."
+    )
+
+    def validate(self, password, user=None):
+
+        has_lower = any(char for char in password if char.islower())
+        has_upper = any(char for char in password if char.isupper())
+        has_digit = any(char for char in password if char.isdigit())
+        has_special_char = any(char for char in password if char in self.SPECIAL_CHARS)
```

Comment:
Le `any` me confuse un peu, normalement c'est pour tester qu'au moins un des éléments de la liste est `True.`.

Ici c'est plutôt un `len(...) > 0` que j'aurai fait. 

Ou encore mieux `any([char.isupper() for char in password]`

Reply:
Le `any` est utilisé comme un *generator comprehension*, il s'arrête à la première expression *truthy* comparé à un `len` qui va tout évaluer.

Reply:
Dans ce cas un `any(char.isupper() for char in password)` serait quand même plus élégant et aussi performant

Reply:
Yes, je prends ce feedback aussi !